### DISCUSSION TITLE
INSFVMixingLengthReynoldsStress explanation

### DISCUSSION THREAD

**Discussion by aiskhak** [idaholab/moose#22989] [Q&A Modules: Navier-Stokes] [answered]
Hi All,

Please help me in understanding of the INSFVMixingLengthReynoldsStress kernel.

I have the following questions:

1) Variable `_axis_index` represents the momentum equation component for which we are using the kernel (`x`, `y`, or `z`). What does then variable `_index` denote?

2) Let's say, `_axis_index=0` (i.e., we are solving for the `x` component) and the problem is 3D. 

Then we first compute the following quantity:

`norm_strain_rate = grad_u(0) * _normal(0) + grad_v(0) * _normal(1) + grad_w(0) * _normal(2)`

i.e.,

`norm_strain_rate = du/dx * nx + dv/dx * ny + dw/dx * nz`

And second we add:

```
const ADRealVectorValue & var_grad = _index == 0 ? grad_u : (_index == 1 ? grad_v : grad_w);
norm_strain_rate += var_grad * _normal;
```

It is not clear for me what is the result of these 2 operations? Seems that after the first one we are getting a vector, and then add to it tensor?

3) What is going on in this part of the code?

```
  if (_face_type == FaceInfo::VarFaceNeighbors::ELEM ||
      _face_type == FaceInfo::VarFaceNeighbors::BOTH)
  {
    const auto dof_number = _face_info->elem().dof_number(_sys.number(), _var.number(), 0);
    // norm_strain_rate is a linear combination of degrees of freedom so it's safe to straight-up
    // index into the derivatives vector at the dof we care about
    _ae = norm_strain_rate.derivatives()[dof_number];
    _ae *= -rho * eddy_diff;
  }
  if (_face_type == FaceInfo::VarFaceNeighbors::NEIGHBOR ||
      _face_type == FaceInfo::VarFaceNeighbors::BOTH)
  {
    const auto dof_number = _face_info->neighbor().dof_number(_sys.number(), _var.number(), 0);
    _an = norm_strain_rate.derivatives()[dof_number];
    _an *= rho * eddy_diff;
  }
```

4) What is going on in this part of the code?

```
void
INSFVMixingLengthReynoldsStress::gatherRCData(const FaceInfo & fi)
{
  if (skipForBoundary(fi))
    return;

  _face_info = &fi;
  _normal = fi.normal();
  _face_type = fi.faceType(_var.name());

  processResidualAndJacobian(computeStrongResidual() * (fi.faceArea() * fi.faceCoord()));

  if (_face_type == FaceInfo::VarFaceNeighbors::ELEM ||
      _face_type == FaceInfo::VarFaceNeighbors::BOTH)
    _rc_uo.addToA(&fi.elem(), _index, _ae * (fi.faceArea() * fi.faceCoord()));
  if (_face_type == FaceInfo::VarFaceNeighbors::NEIGHBOR ||
      _face_type == FaceInfo::VarFaceNeighbors::BOTH)
    _rc_uo.addToA(fi.neighborPtr(), _index, _an * (fi.faceArea() * fi.faceCoord()));
}
```

5) The kernel returns "the force due to the Reynolds stress term":

`-1 * rho * eddy_diff * norm_strain_rate;`

what is the structure of this term, I mean if it is a tensor, what is its dimension?

**Comment by GiudGiud** [comment, marked as answer]
Hello

1)
For an axis, you would expect 'x', 'y', 'z'. The axis_index is simply 0, 1, 2 for those axes respectively. 

2)
var_grad * normal is the dot product of two vectors, so the output is an (AD)Real. 
We are adding to the normalized strain rate, a number

3)
In this part of the code we are tallying the contribution of the face flux to the Rhie Chow coefficient of the element (ae) and the neighbor (an) of the face.

4) 
This part of the code is moving this tallied contribution into the actual coefficient, held by the RC Interpolator user object. 
The contribution is weighted by the area of the face (coord() is the RZ radius)

5)
This term is a scalar, an (AD)Real. It s the contribution to the residual of the momentum equation of this term only for one axis. 

Hope this helps,
Guillaume

**Reply by aiskhak** [reply]
Thank you for the answer. Let me make sure I understand everything:

1. For the given `_axis_index` (`x`, `y`, or `z`) specified in the input file, this kernel is called only once for the corresponding `_index` (`0`, `1`, or `2`), right? 

In other words, 

if `_axis_index=x`, then by default `_index=0`
if `_axis_index=y`, then by default `_index=1`
if `_axis_index=z`, then by default `_index=2`

right?

2. Let's then assume `dim=3`, `_axis_index=x`, `_index=0` , then the kernel will return:

`norm_strain_rate = grad_u(0) * _normal(0) + grad_v(0) * _normal(1) + grad_w(0) * _normal(2) + grad_u * _normal` 

which is

`norm_strain_rate = du/dx*nx + dv/dx*ny + dw/dx*nz + du/dx*nx + du/dy*ny + du/dz*nz`

If `_axis_index=y`, `_index=1`:

`norm_strain_rate = grad_u(1) * _normal(0) + grad_v(1) * _normal(1) + grad_w(1) * _normal(2) + grad_v * _normal` 

which is

`norm_strain_rate = du/dy*nx + dv/dy*ny + dw/dy*nz + dv/dx*nx + dv/dy*ny + dv/dz*nz`

If `_axis_index=z`, `_index=2`:

`norm_strain_rate = grad_u(2) * _normal(0) + grad_v(2) * _normal(1) + grad_w(2) * _normal(2) + grad_w * _normal` 

which is

`norm_strain_rate = du/dz*nx + dv/dz*ny + dw/dz*nz + dw/dx*nx + dw/dy*ny + dw/dz*nz`

right?

2a. We do not take the derivatives of the above because we are in the weak formulation, correct?

3. Gotcha.

4. Gotcha.

5. Gotcha.

Thank you again.

**Reply by GiudGiud** [reply]
1) yes. The mapping is as expected, even if the variable names seem unclear right now

2)
This looks correct.
It's (grad_vel + t(grad_vel)) * n but taken only on one axis

2a.
No this is a strong form we re doing finite volume.
However, this is a face/flux kernel, which allows us to drop the derivative.
It's because we are using the divergence theorem to compute this term.

**Reply by aiskhak** [reply]
ok, gotcha. Thanks! I may follow up with the items 3 and 4 later.

**Comment by aiskhak** [comment]
Can you also explain - how the kernel works for cylindrical coordinates? At which point the code knows that the equations are in the cylindrical coordinates and adjusts the computed Reynolds stresses?
